File: algorithms/java/src/test/java/leetcode/LC318MaximumProductOfWordLengths.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/maximum-product-of-word-lengths/
 * #medium
 */
public class LC318MaximumProductOfWordLengths {
    public class Solution {
        public int maxProduct(String[] words) {
            int[] signature = new int[words.length];
            for (int i = 0; i < words.length; i++) {
                for (char character : words[i].toCharArray()) {
                    signature[i] |= 1 << (character - 'a');
                }
            }
            int maximum = 0;
            for (int i = 0; i < words.length - 1; i++) {
                for (int j = i + 1; j < words.length; j++) {
                    if ((signature[i] & signature[j]) == 0) {
                        maximum = Math.max(maximum, words[i].length() * words[j].length());
                    }
                }
            }
            return maximum;
        }
    }

    @Test
    public void test_example_1() throws Exception {
        String[] input = {"abcw", "baz", "foo", "bar", "xtfn", "abcdef"};
        assertEquals(16, new Solution().maxProduct(input));
    }

    @Test
    public void test_example_2() throws Exception {
        String[] input = {"a", "ab", "abc", "d", "cd", "bcd", "abcd"};
        assertEquals(4, new Solution().maxProduct(input));
    }

    @Test
    public void test_example_3() throws Exception {
        String[] input = {"a", "aa", "aaa", "aaaa"};
        assertEquals(0, new Solution().maxProduct(input));
    }
}

File: algorithms/code/leetcode/lc318_maximum_product_of_word_lengths/lc318_maximum_product_of_word_lengths.py
# https://leetcode.com/problems/maximum-product-of-word-lengths/

import unittest


class Solution:
    def maxProduct(self, words: list[str]) -> int:
        signature = [0] * len(words)
        for i, _ in enumerate(words):
            for character in words[i]:
                signature[i] |= 1 << (ord(character) - ord('a'))
        maximum = 0
        for i in range(len(words) - 1):
            for j in range(i + 1, len(words)):
                if (signature[i] & signature[j]) == 0:
                    maximum = max(maximum, len(words[i]) * len(words[j]))
        return maximum


class TestCode(unittest.TestCase):
    def test_example_1(self) -> None:
        words = ['abcw', 'baz', 'foo', 'bar', 'xtfn', 'abcdef']
        assert Solution().maxProduct(words) == 16

    def test_example_2(self) -> None:
        words = ['a', 'ab', 'abc', 'd', 'cd', 'bcd', 'abcd']
        assert Solution().maxProduct(words) == 4

    def test_example_3(self) -> None:
        words = ['a', 'aa', 'aaa', 'aaaa']
        assert Solution().maxProduct(words) == 0
